src/Interpreters/JoinInfo.h

```diff
@@ -134,6 +148,30 @@ struct JoinCondition
     /// Unlike the join predicates, these conditions can be arbitrary expressions.
     std::vector<JoinActionRef> residual_conditions;
 
+    void fixReferences(const JoinExpressionActions & expression_actions)
+    {
+        for (auto & predicate : predicates)
+        {
+            predicate.left_node = JoinActionRef(&expression_actions.left_pre_join_actions->findInOutputs(predicate.left_node.getColumnName()), expression_actions.left_pre_join_actions.get());
```

Comment:
I think it would be better to get a `map : Node * -> Node *` from `ActionsDAG::clone` (and `JoinExpressionActions::clone`), and replace references based on it.